pkg/dataobj/planner/logical/format_tree_test.go

```diff
@@ -0,0 +1,103 @@
+package logical
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/grafana/loki/v3/pkg/dataobj/internal/metadata/datasetmd"
+	"github.com/grafana/loki/v3/pkg/dataobj/planner/schema"
+)
+
+type testDataSource struct {
+	name   string
+	schema schema.Schema
+}
+
+func (t *testDataSource) Schema() schema.Schema { return t.schema }
+func (t *testDataSource) Name() string          { return t.name }
+
+func TestFormatSimpleQuery(t *testing.T) {
+	// Build a simple query plan:
+	// SELECT id, name FROM users WHERE age > 21
+	ds := &testDataSource{
+		name: "users",
+		schema: schema.Schema{
+			Columns: []schema.ColumnSchema{
+				{Name: "id", Type: datasetmd.VALUE_TYPE_UINT64},
+				{Name: "name", Type: datasetmd.VALUE_TYPE_STRING},
+				{Name: "age", Type: datasetmd.VALUE_TYPE_UINT64},
+			},
+		},
+	}
+
+	scan := NewScan(ds.Name(), ds.Schema())
+	filter := NewFilter(scan, Gt("age_gt_21", Col("age"), LitI64(21)))
+	proj := NewProjection(filter, []Expr{Col("id"), Col("name")})
+
+	f := newTreeFormatter()
+
+	expected := `
+Projection id=VALUE_TYPE_UINT64 name=VALUE_TYPE_STRING
+├── Column #id
+├── Column #name
+└── Filter expr=age_gt_21
+    ├── BinaryOp op=(>) name=age_gt_21
+    │   ├── Column #age
+    │   └── Literal value=21 type=VALUE_TYPE_INT64
+    └── MakeTable name=users`
+
+	require.Equal(t, expected, "\n"+f.Format(proj))
+}
+
+func TestFormatDataFrameQuery(t *testing.T) {
+	// Calculate the sum of sales per region for the year 2020
+	ds := &testDataSource{
+		name: "orders",
+		schema: schema.Schema{
+			Columns: []schema.ColumnSchema{
+				{Name: "region", Type: datasetmd.VALUE_TYPE_STRING},
+				{Name: "sales", Type: datasetmd.VALUE_TYPE_UINT64},
+				{Name: "year", Type: datasetmd.VALUE_TYPE_UINT64},
+			},
+		},
+	}
+
+	df := NewDataFrame(
+		NewScan(ds.Name(), ds.Schema()),
+	).Filter(
+		Eq("year_2020", Col("year"), LitI64(2020)),
+	).Project(
+		[]Expr{
+			Col("region"),
+			Col("sales"),
+			Col("year"),
+		},
+	).Aggregate(
+		[]Expr{Col("region")},
+		[]AggregateExpr{
+			Sum("total_sales", Col("sales")),
+		},
+	)
+
+	f := newTreeFormatter()
+
+	expected := `
+Aggregate groupings=(region) aggregates=(total_sales)
+├── GroupExprs
+│   └── Column #region
+├── AggregateExprs
+│   └── AggregateExpr op=sum
```

Comment:
nit: I would have expected the the AggregateExpr would also be printed with its name, like BinaryOp:

```suggestion
│   └── AggregateExpr op=sum name=total_sales
```